### ISSUE TITLE
Batch/iteratively build dataset class and train model for big data

### ISSUE THREAD

**Issue by fmcquarrie3** [lyst/lightfm#611]
My question builds off [Issue 234](https://github.com/lyst/lightfm/issues/234). 

I have ~1.4M users, 10 items, 623 user features, no item features, and each user has interacted with only one item. Based on the documentation, my pipeline for building a LightFM recommender would be:
```
from lightfm.data import Dataset
from lightfm import LightFM

userIDs = ['a', 'b', 'c', ...]
itemIDs = [0, 1, 2, .., 9]

# all possible values of all possible features
all_unique_user_features_formatted = ['cat_feature1:unique_value1', 'cat_feature1:unique_value2', ...,  'cont_feature2_name', 'cont_feature3_name', ...]
# which user has which feature
user_features_dense_formatted = [('a', {'cat_feature1:valuea1': 1, 'cont_feature2_name': num_value_of_feat, ...}), ('b', {'cat_feature1:valueb1': 1, 'cont_feature2_name': num_value_of_feat, ...}), ...]
interactions_dense = [('a', 3), ('b', 1), ('c', 0), ...]

# format data into sparse structure with Dataset class
dataset = Dataset()
dataset.fit(users = userIDs, 
            items = itemIDs, 
            user_features = all_unique_user_features_formatted)
            
user_features_sparse = dataset.build_user_features(user_features_dense_formatted)
interactions_sparse, weights = dataset.build_interactions(interactions_dense)

# Fit the LightFM model
my_model = LightFM()
my_model.fit(interactions_sparse, 
            sample_weights = weights, 
            user_features = user_features_sparse)
```
The issue: the `interactions_dense` iterable can fit in local memory, but `user_features_dense_formatted` cannot (I preprocessed all the data into the correct format with `pyspark`, and hoped to call `pyspark_df.toPandas().to_numpy()` to get the simple iterable required by the `Dataset` class, but alas it is too big). 

Note: I had orginally formatted all categorical and continuous variables in the string format "feature_name:feature_value" and using the list input to build based off the advice from [this post](https://towardsdatascience.com/how-i-would-explain-building-lightfm-hybrid-recommenders-to-a-5-year-old-b6ee18571309). When we first encountered the memory issue, we thought it was due to size of storing many strings. We then switched to using the dictionary formatted input to build, following [Issue 433](https://github.com/lyst/lightfm/issues/433#issuecomment-472888625) and [Issue 536](https://github.com/lyst/lightfm/issues/536), so that the continuous variables take up less space. But this still creates memory issues.

In [Issue 234](https://github.com/lyst/lightfm/issues/234), you suggested reading data from the disk in batches, and making subsequent `my_model.fit_partial(...)` calls. However, a clarification:

Q1: Will calling `my_model.fit_partial()` multiple times with the same interaction data but different sets of user features have the same effect? In other words, will
```
# format data into sparse structure with Dataset class
dataset = Dataset()
dataset.fit(users = userIDs, 
            items = itemIDs, 
            user_features = all_unique_user_features_formatted)
            
interactions_sparse, weights = dataset.build_interactions(interactions_dense) 
# build first batch of user features
user_features_sparse1 = dataset.build_user_features(user_features_dense_formatted1)
# build second batch of user features
user_features_sparse2 = dataset.build_user_features(user_features_dense_formatted2)

# Fit the LightFM model in two batches
my_model = LightFM()
my_model.fit_partial(interactions_sparse, 
            sample_weights = weights, 
            user_features = user_features_sparse1)
my_model.fit_partial(interactions_sparse, 
            sample_weights = weights, 
            user_features = user_features_sparse2)
```
have the effect of incorporating all interactions and user features, just in batches that fit in memory? `user_features_dense_formatted1` and `user_features_dense_formatted2` would contain information about the same features, but mutually exclusive sets of users. I ask because in the following toy example:
```
# setup toy data
userIds = ['a', 'b', 'c', 'd', 'e']
itemIds = [0, 1, 2]
all_feat_values = ['hi', 'bye', 'maybe']

user_features_raw = [('a', ['hi', 'bye']), ('b', ['maybe', 'bye']), ('c', ['hi']), ('d', ['bye']), ('e', ['hi', 'bye', 'maybe'])]

user_features1 = user_features_raw[0:2]
user_features2 = user_features_raw[2:5]

# build format
dataset = Dataset()
dataset.fit(users = userIds, items = itemIDs, user_features = all_feat_values)
user_features1_sparse = toy.build_user_features(user_features1)
user_features2_sparse = toy.build_user_features(user_features2)
interactions_dense = zip(userIds, [0, 2, 1, 1, 2])
interactions_sparse, weights = toy.build_interactions(interactions_dense)

# fit model
toy_model = LightFM()
toy_model.fit_partial(interactions = interactions_sparse, user_features = user_features1_sparse)
u1 = toy_model.get_user_representations()
toy_model.fit_partial(interactions = interactions_sparse, user_features = user_features2_sparse)
u2 = toy_model.get_user_representations()

# did fitting the second time make a difference?
(u1[0] == u2[0]).all() #evals to true
(u1[1] == u2[1]).all() #evals to true
```
fitting the model twice with the same interactions but non-overlapping sets of user features seems to only incorporate the first fitting. 

Q2: Is there some way to confirm that `fit_partial` incorporated the new information? 

tl;dr: How do you recommend building inputs & fitting models when the user_features iterable can't fit in local memory but the interaction data can?

Appreciate any insight you could give, thanks!

**Comment by fmcquarrie3** [comment]
Eventually gave up on constructing model inputs using the `Dataset` class. Instead, we constructed the `scipy.sparse` input matrices by hand, using the `scipy.sparse.coo_matrix((data, (i, j)), shape = (n,m))` constructor. We were able to fit the data & coordinate arrays into memory after adjusting our data and node memory capacity. If you need to construct the `scipy` inputs by hand but cannot fit the coordinates in memory, you can potentially explore something like
```
matrix = data.rdd \
                      .mapPartitions(fn_to_construct_mini_scipy_sparse_matrix) \
                      .treeReduce(lambda x, y: scipy.sparse.vstack([x, y], "coo"))
```

Still curious if anyone knows the answer to my original question but closing the issue since original need is resolved.
